src/sieve/sampler_basic.h

```diff
@@ -0,0 +1,99 @@
+/* "Trapdoors for Hard Lattices and New Cryptographic Constructions"
+   Craig Gentry, Chris Peikert, Vinod Vaikuntanathan */
+
+#ifndef FPLLL_SIEVE_SAMPLER_BASIC_H
+#define FPLLL_SIEVE_SAMPLER_BASIC_H
+
+#include "../fplll.h"
+#include "../gso.cpp"
+
+
+using namespace std;
+using namespace fplll;
+
+template<class ZT, class F>
+class KleinSampler {
+
+  public:
+  bool verbose;
+
+  KleinSampler (ZZ_mat<ZT>& B, bool verbose);
+  ~KleinSampler();
+  void print_param ();
+  Z_NR<ZT> sampleZ(F c, F s);
+  Z_NR<ZT> sampleZ_basic(F c, F s);
+  void set_verbose (bool verbose);
+
+  NumVect<Z_NR<ZT> > sample();
+  
+  private:
+
+  /**
+   * GSO object of B
+   */
+  int nr, nc;
+  double logn2;
+  gmp_randstate_t state;
+  ZZ_mat<ZT> b, u, uInv;
+  MatGSO<Z_NR<ZT>, F> *pGSO;
```

Comment:
My C++ isn't strong enough but couldn't this just be a MatGSO object (not a pointer)?